File: python/p004.py
def is_palindrome(n):
    """Checks if n is a palindromic number."""
    s = str(n)
    return s == s[::-1]


def p004(N):
    """Finds the largest palindrome made from the product
    of two N-digit numbers."""
    Ndigits = list(range(10**N-1, 10**(N-1), -1))
    max_palindrome = 0
    for i in Ndigits:
        for j in Ndigits:
            if is_palindrome(i*j):
                max_palindrome = max(max_palindrome, i*j)
    return max_palindrome


if __name__ == "__main__":
    print(p004(3))

File: julia/p004.jl
function is_palindrome(n::Int)
    s = string(n)
    parity = length(s) % 2
    if s[1:end÷2] == s[end:-1:end÷2+1+parity]
        return true
    else
        return false
    end
end

function p004(n::Int)
    largest = 1
    for (i, j) in Iterators.product(1:n-1, 1:n-1)
        if is_palindrome(i * j) && i * j > largest
            largest = i * j
        end
    end
    return largest
end

println(p004(1000))
